common/src/main/java/com/wynntils/services/mapdata/attributes/FullCategoryAttributes.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.services.mapdata.attributes;
+
+import com.wynntils.core.components.Services;
+import com.wynntils.services.mapdata.attributes.type.MapAttributes;
+import com.wynntils.services.mapdata.type.MapCategory;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+public class FullCategoryAttributes extends DerivedAttributes {
+    private final String categoryId;
+
+    public FullCategoryAttributes(String categoryId) {
+        this.categoryId = categoryId;
+    }
+
+    protected <T> T getAttribute(Function<MapAttributes, T> getter) {
+        for (String id = categoryId; id != null; id = getParentCategoryId(id)) {
+            Stream<MapAttributes> allAttributes = Services.MapData.getCategoryDefinitions(id)
+                    .map(MapCategory::getAttributes)
+                    .filter(Objects::nonNull);
+            Optional<T> attribute =
+                    allAttributes.map(getter).filter(Objects::nonNull).findFirst();
+            if (attribute.isPresent() && !(attribute.get() instanceof Integer i && i == 0)) {
```

Comment:
I really really dislike API:s that use Integer or any other of the boxed value classes. That typically always end up of breaking the principle of least surprise, since now you can suddenly have null for what you thought was an integer. I have explicitly tried to make sure that the API as provided by the MapAttributes should have proper `int` for the (two) places where it returns numerical values.

Reply:
Okay, it was just an idea. I too dislike using boxed classes, but Optional, ErrorOr or any other wrapper class is too explicit for this kind of data structure. Perhaps this is where Kotlin's "nullability" safety would help..